### PR TITLE
BUG: optimize.minimize: fix `trust-constr` `status=2` despite constraint

### LINKED ISSUE #18882: minimize does not satisfy nonlinear constraint even if keep_feasible=True
### Describe your issue.

I am creating a general app that can calculate the probability of failure of engineering systems based on a limit state function.  In its most simple form this requires the solution of a minimization problem with an equality constraint.  The app works on almost all tests but have found that one specifically (very academic) causes problems where the equality constraint is not satisfied. 

There is no error message but the result is not correct.  The result is close to x = [0,0] but it should be x = [0, 4].  The resulting equality constraint has a value equal to 1.0 and not 0.0 as required.  The correct result can be obtained if u0 is set to other values like [0,1] but not when set to [1,0].  My expectation was that the equality constraint would be always satisfied if the minimization was successful but it appears it is not, even when keep_feasible=True.  I can solve the issue by setting u0 to [0,1] but the goal of the app I am developing is to work for a general set of limit state functions (constraint functions) and so would like to set u0 = zeros for all cases.

I know this is a challenging problem because the gradient of both the objective and constraint functions is zero at u0 = [0,0].  But why does minimize also fail to satisfy the equality constraint when u0 = [1,0]?

I am using method="trust-constr" because my general app also requires bounds (linear constraints) on the variables and the objective and constraint functions can return gradient and hessian to speed up the solution in large problems.

### Reproducing Code Example

```python
import numpy as np
from scipy import optimize
from IPython.core.interactiveshell import InteractiveShell

def lsf(u):
    u1, u2 = u

    a, b, c, d = [0.0, 0.0, 3.0, 4.0]

    g = 1.0 + (((u1 - a) ** 2) / c**2 - (u2 - b) ** 2 / d**2)

    return g

def of(u):
    d = np.sum(u**2) # sum of squares
    return d

cons = optimize.NonlinearConstraint(lsf, 0, 0, keep_feasible=True)
u0 = [0.0, 0.0]

res = optimize.minimize(
            of,
            u0,
            method="trust-constr",
            constraints=cons,
        )

res
```

### Error message

```shell

```

### SciPy/NumPy/Python version and system information

```shell
1.11.1 1.25.1 sys.version_info(major=3, minor=11, micro=4, releaselevel='final', serial=0)
{
  "Compilers": {
    "c": {
      "name": "clang",
      "linker": "ld64",
      "version": "15.0.7",
      "commands": "arm64-apple-darwin20.0.0-clang"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "0.29.35",
      "commands": "cython"
    },
    "c++": {
      "name": "clang",
      "linker": "ld64",
      "version": "15.0.7",
      "commands": "arm64-apple-darwin20.0.0-clang++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld64",
      "version": "12.2.0",
      "commands": "/Users/runner/miniforge3/conda-bld/scipy-split_1687995684757/_build_env/bin/arm64-apple-darwin20.0.0-gfortran"
    },
    "pythran": {
      "version": "0.13.1",
      "include directory": "../../../_build_env/venv/lib/python3.11/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "arm64",
      "family": "aarch64",
      "endian": "little",
      "system": "darwin"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "darwin"
    },
    "cross-compiled": true
  },
  "Build Dependencies": {
    "blas": {
      "name": "blas",
      "found": true,
      "version": "3.9.0",
      "detection method": "pkgconfig",
      "include directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/include",
      "lib directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/lib",
      "openblas configuration": "unknown",
      "pc file directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/lib/pkgconfig"
    },
    "lapack": {
      "name": "lapack",
      "found": true,
      "version": "3.9.0",
      "detection method": "pkgconfig",
      "include directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/include",
      "lib directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/lib",
      "openblas configuration": "unknown",
      "pc file directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/lib/pkgconfig"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.10.4",
      "detection method": "pkgconfig",
      "include directory": "/Users/Carrasco/miniconda3/envs/carrasco_env311/include"
    }
  },
  "Python Information": {
    "path": "/Users/Carrasco/miniconda3/envs/carrasco_env311/bin/python",
    "version": "3.11"
  }
}
```

### PR DESCRIPTION
**Pull request by ellieLitwack**
#### Reference issue

Fixes #18882.
#### What does this implement/fix?

Previously, trust-constr returned status=2 ("`xtol` termination condition is satisfied.") even if the constraint was not satisfied. Now, trust-constr will continue to iterate if the constraint is not satisfied. Also, this adds a test which checks for this issue in the scenario in #18882.

### CODE REVIEW

**Comment by ellieLitwack** [comment]
This is also related to #10112.

File: scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py [hunk 1]
```diff
@@ -131,7 +131,9 @@ def _minimize_trustregion_constr(fun, x0, args, grad,
         Tolerance for termination by the norm of the Lagrangian gradient.
         The algorithm will terminate when both the infinity norm (i.e., max
         abs value) of the Lagrangian gradient and the constraint violation
-        are smaller than ``gtol``. Default is 1e-8.
+        are smaller than ``gtol``. Also the tolerance for the maximum
+        constraint violation at the solution if an equality constraint is
```
Comment (andyfaff) [inline scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:135]: `constr_violation` is the maximum constraint violation for both inequality and equality constraints.
I don't understand why this last sentence has been added, it's probably not necessary.

File: scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py [hunk 2]
```diff
@@ -439,7 +441,7 @@ def stop_criteria(state, x, last_iteration_failed,
                     return True
             if state.optimality < gtol and state.constr_violation < gtol:
                 state.status = 1
-            elif state.tr_radius < xtol:
+            elif state.tr_radius < xtol and state.constr_violation < gtol:
```
Comment (andyfaff) [inline scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:444]: By adding this extra stopping condition the docstring for `xtol` becomes incorrect.

The docstring should read something like:

> Tolerance for termination by the change of the independent variable.
        The algorithm will terminate when ``tr_radius < xtol and constr_violation < gtol``, where
        ``tr_radius`` is the radius of the trust region used in the algorithm.
        Default is 1e-8.


In any case, I'm not sure if this is the desired behaviour. I'd prefer to keep the stopping condition unchanged (i.e. revert this line). Instead should we add an extra item to `TERMINATION_MESSAGES`, "constraints are not satisfied", and change L538 to:

```python
if state.constr_violation > gtol:
    result.status = 4    # TERMINATION_MESSAGES[4] = "constraints are not satisfied"
result.success = True if result.status in (1, 2) else False
```

My rationale is that it's no change to the stopping condition, but still informs the user that something has gone wrong. It's relatively common that the constraints are always infeasible, in which case the proposed PR would keep going until maximum iterations had been reached. That's undesirable as it artificially prolongs minimisation.

I don't think adding the extra condition in L442 would actually improve a minimisation anyway. I'm ready to be convinced with evidence though.

File: scipy/optimize/tests/test_minimize_constrained.py [hunk 3]
```diff
@@ -614,6 +614,44 @@ def callback(x, info):
         # Also check existence of the 'niter' attribute, for backward
         # compatibility
         assert_(result.get('niter', -1) == 1)
+    
+    def test_issue_18882(self):
+        # https://github.com/scipy/scipy/issues/18882
+        # Test that success is false if the constraint is not met.
```
Comment (andyfaff) [inline scipy/optimize/tests/test_minimize_constrained.py:620]: this test does not actually test that `res.success is False` if the constraints are not met.
The test is also a lot longer than it needs to be.

**Review by andyfaff — CHANGES_REQUESTED** [review]
@ellieLitwack, thank you for preparing this PR. I think we should take a slightly different approach, as suggested in the original issue.

**Comment by lucascolley** [comment]
Hi @ellieLitwack, are you interested in returning to this PR?

**Comment by tylerjereddy** [comment]
This PR has blocking comments, including at the fundamental design level, and no activity for the past 6 months, so I will bump the release milestone for now.

**Comment by CarrascoCesar** [comment]
Just wanted to revive the discussion related to this issue.  Should  TERMINATION_MESSAGES be expanded to include a status 4 indicating that the constraints have not been satisfied?

**Comment by CarrascoCesar** [comment]
Hello @andyfaff do you know if this issue has beed addressed through some other PR?

**Comment by lucascolley** [comment]
> do you know if this issue has beed addressed through some other PR?

I don't think it has - would you like to take over the PR? Feel free to open a separate PR, and [add any coauthors](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) if you reuse code from this PR.

**Comment by CarrascoCesar** [comment]
Thanks @lucascolley this will be my first PR but I'm up for it.  Will definitely reference the comments and code in this PR.

**Comment by lucascolley** [comment]
Great! Our contributor guidance is at https://scipy.github.io/devdocs/dev/index.html, feel free to give me a shout if you need help

**Comment by CarrascoCesar** [comment]
Hello @lucascolley I created the PR https://github.com/scipy/scipy/pull/21407 but it has failed 2 checks that I think are unrelated to the changes I made.  What should be my next step?  Should I wait for Andy to finish the review?

**Comment by lucascolley** [comment]
Thanks @CarrascoCesar - please just wait for someone to take a look! I can try to take a preliminary look in the coming days.

**Comment by j-bowhay** [comment]
Closing as superceded by #21407 but thanks for getting the ball rolling @ellieLitwack

**Pull request closed without merging**